### ISSUE TITLE
Split rendered image into smaller pieces using a <table> element?

### ISSUE THREAD

**Issue by danielnez1** [tenox7/wrp#90]
Hello,

I have been trying WRP to load pages on the browsers supplied with the newer versions of PC/GEOS (Webmagik on Breadbox Ensemble and Skipper on Newdeal Office 3.2a). These environments can run into out of memory and speed issues when loading large images (I suspect this is due to them being DOS based real mode GUI environments relying on paging to access memory beyond the 640k barrier).

With that in mind, i have been experimenting with splitting up the rendered image of a website into smaller pieces on a static HTML file using a table, with each chunk being placed inside a <td> element. based on my anecdotal observations, it does seem to render the image elements quicker compared to rendering a single image, however would such an approach break the image map approach WRP uses to track mouse clicks (I am not too familiar with how they work)? 

Cheers,

Dan

**Comment by tenox7** [comment]
Ha! That's a neat idea. I think you don't even need a table you could potentially just add images beside each other in a row and <BR> for a new row. As for breaking image maps, you could just have a separate map for each image.

**Comment by pomosapiens** [comment]
Going further, if we think of WRP as VNC, but just inside a browser, then you could think about pasts of a canvas being damaged and needing a redraw, like VNC and X11 do.

You'd cut the image into the regions that might need to change independently.  The regions won't be the same size; they might be like an octree.  Assuming the vintage browser can only request a whole page at a time (no fetching pieces async), you still send the whole page but make sure the client can still hit its cache for unchanged imagemaps.  Then, with luck the client only needs to download the changed piece.

**Comment by rhaleblian** [comment]
Going further, if we think of WRP as VNC, but just inside a browser, then you could think about parts of an image canvas (the rendered page) being damaged and needing a redraw, like VNC and X11 do.

You'd cut the image into the regions that might need to change independently. The regions won't be the same size; they might be like an octree. Assuming the vintage browser can only request a whole page at a time (no fetching pieces async), you still send the whole page but make sure the client can still hit its cache for unchanged imagemaps. Then, with luck the client only needs to download the changed piece.

**Comment by tenox7** [comment]
Right I can see that working. You could (perhaps) control cache of individual images/slices and basically expire images you want to update and download for the changed ones. I think some experimentation is in order.

**Comment by tenox7** [comment]
I believe even in the 90s browsers still used HEAD (even in HTTP/1.0 https://datatracker.ietf.org/doc/html/rfc1945#section-8.2) to check if resource (image in this case) has been modified since last request. Of course browsers vary a lot and testing is required, but theoretically the idea is solid. I will probably write a test tool that will output a grid of small images that will be constantly changing and we can get it tested on all different browsers.

**Comment by danielnez1** [comment]
@rhaleblian Sorry for the late reply, no I never got around to implementing it :-(

As others have  pointed out, I discovered my rudimentary experiments (I never got as far as hacking the Go code) showed that the different browsers handle the visual formatting of <table>, <tr> and <td> elements slightly different, so cause whitespace issues.

**Comment by tenox7** [comment]
Right! Before CSS there was no way to control this and different browsers handled this differently. 

There are other options I'm exploring in the mean time.

GIF encoding can be speed up dramatically by using concurrency.
Perhaps heavier compression, less colors could be used by default, especially for less graphical pages.
GIF image download could start earlier on the client and be blocked or even made progressive as it's being encoded.

**Comment by rhaleblian** [comment]
Good research!  Such were the days.

**Comment by maagenta** [comment]
I don't think it will improve the speed. More images implies more HTTP requests. For an old computer, each HTTP request takes time. I was trying to do something like that (dividing the webview into strips) in this project https://github.com/maagenta/webone-snapshot-viewer/blob/main/ARCHITECTURE.md and it works great, but it has the disadvantage that more requests create more pressure on old systems.

I just saw this project and I think it's perfect the way it works. Stripping the website would only work well for newer systems. So, in that case, I think the best solution is to create a flag that activates that mode.
